File: src/source/Yaaf.Database/NUnitInitializer.cs
// ----------------------------------------------------------------------------
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.
// ----------------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Data.Entity;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Yaaf.Database {

	public class MyDropCreateDatabaseIfModelChanges<TContext> : DropCreateDatabaseIfModelChanges<TContext> where TContext : DbContext {
		private Action<TContext> seedMethod;
		public MyDropCreateDatabaseIfModelChanges (Action<TContext> seedMethod)
		{
			this.seedMethod = seedMethod;
		}
		protected override void Seed (TContext context)
		{
			base.Seed (context);
			if (seedMethod != null) {
				seedMethod (context);
			}
		}
	}
	public class MyCreateDatabaseIfNotExists<TContext> : CreateDatabaseIfNotExists<TContext> where TContext : DbContext {
		private Action<TContext> seedMethod;
		public MyCreateDatabaseIfNotExists (Action<TContext> seedMethod)
		{
			this.seedMethod = seedMethod;
		}
		protected override void Seed (TContext context)
		{
			base.Seed (context);
			if (seedMethod != null) {
				seedMethod (context);
			}
		}
	}


	public class NUnitInitializer<TContext> : IDatabaseInitializer<TContext> where TContext : DbContext {

		private DropCreateDatabaseIfModelChanges<TContext> changer;
		private CreateDatabaseIfNotExists<TContext> creator;
		public NUnitInitializer ()
		{
			changer = new MyDropCreateDatabaseIfModelChanges<TContext> (this.Seed);
			creator = new MyCreateDatabaseIfNotExists<TContext> (this.Seed);
		}

		protected virtual void Seed (TContext context)
		{
		}

		public void InitializeDatabase (TContext context)
		{
			try {
				changer.InitializeDatabase (context);
			} catch (NotSupportedException) {
				// Database not existent
				creator.InitializeDatabase (context);
			}
		}
	}

}

File: src/source/Yaaf.Database/NUnitInitializer.fs
﻿// ----------------------------------------------------------------------------
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.
// ----------------------------------------------------------------------------
namespace Yaaf.Database
open System
open System.Collections.Generic
open System.Data.Entity
open System.Linq
open System.Text
open System.Threading.Tasks

type MyDropCreateDatabaseIfModelChanges<'TContext when 'TContext :> DbContext > (seedMethod : 'TContext -> unit) =
    inherit DropCreateDatabaseIfModelChanges<'TContext>()
    
    default x.Seed (context:'TContext) =
        base.Seed (context);
        seedMethod(context)

type MyCreateDatabaseIfNotExists<'TContext when 'TContext :> DbContext > (seedMethod : 'TContext -> unit) =
    inherit CreateDatabaseIfNotExists<'TContext>()
    
    override x.Seed (context:'TContext) =
        base.Seed (context)
        seedMethod(context)

type NUnitInitializer<'TContext when 'TContext :> DbContext > () as x =
    let changer = new MyDropCreateDatabaseIfModelChanges<'TContext> (x.Seed);
    let creator = new MyCreateDatabaseIfNotExists<'TContext> (x.Seed);

    abstract Seed : 'TContext -> unit
    default x.Seed (context:'TContext) = ()

    abstract InitializeDatabase : 'TContext -> unit
    default x.InitializeDatabase (context:'TContext) =
        try
            changer.InitializeDatabase (context);
        with
        | :? NotSupportedException ->
            // Database not existent
            creator.InitializeDatabase (context)

    interface IDatabaseInitializer<'TContext> with
      member x.InitializeDatabase c = x.InitializeDatabase c